ui/apps/platform/src/hooks/usePaginatedQuery.ts

```diff
@@ -0,0 +1,187 @@
+}
+
+export type UsePaginatedQueryReturn<Item> = {
+    /** The returned data. Each item in the top level array is a page of data `Item`s */
+    data: Item[][];
+    /** The current page */
+    page: number;
+    /** If the last fetch resulted in an error the `Error` value will appear here */
+    lastFetchError: Error | null;
+    /** If a page fetch is in flight */
+    isFetchingNextPage: boolean;
+    /** If a page fetch is in flight due to the caller refreshing the entire paginated results array */
+    isRefreshingResults: boolean;
+    /** If no more results are present on the server */
+    isEndOfResults: boolean;
+    /** Imperatively request the next page, which will declaratively be available in `data` */
+    fetchNextPage: (immediate?: boolean) => void;
+    /** Clears the entire `data` array and fetches the first page from the beginning */
+    resetPages: () => void;
+    /** Clears the entire `data` array and does not fetch a new page */
+    clearPages: () => void;
+};
+
+/**
+ * This hook is used for infinite loading query patterns via pagination. It's primary use case is a UI list
+ * that we want to infinitely load via something like a "View more" button, that shows all resulting data in
+ * a single UI. i.e. It does not use traditional "paged" UI components.
+ *
+ * In addition to fetching the data, this hook also can handle deduplicating the response in cases where the source
+ * data changes faster than it is loaded in the UI. This is needed to avoid React key conflicts when rendering the
+ * data, as well as for ensuring the results are unique.
+ *
+ * Conceptually it is similar to tanstack-query's https://tanstack.com/query/v4/docs/guides/infinite-queries or
+ * SWR's https://swr.vercel.app/docs/pagination#useswrinfinite so that we can replace this logic if we move
+ * to library based querying in the future.
+ */
+export function usePaginatedQuery<Item, ItemKey>(
+    /** The function used to fetch each page of data */
+    queryFn: (page: number) => Promise<Item[]>,
+    /** The number of items to retrieve per page */
+    pageSize: number,
+    initialOptions: {
+        /** How much debounce delay should occur before the fetch request is sent */
+        debounceRate?: number;
+        /**
+         * The function used to determine the unique key of each item to avoid duplication. If
+         * this parameter is omitted the hook will allow duplicate data in the results.
+         */
+        dedupKeyFn?: (item: Item) => ItemKey;
+        /** Callback function to fire if a request results in an error */
+        onError?: (err: Error) => void;
+        /**
+         *  Whether or not the initial fetch call for the first page should be automatic or require
+         * a manual call to `fetchNextPage`
+         */
+        manualFetch?: boolean;
+    } = {}
+): UsePaginatedQueryReturn<Item> {
+    const [{ debounceRate, dedupKeyFn, onError, manualFetch = false }] = useState(initialOptions);
```

Comment:
Because there is no set function from `useState` hook, is it possible to destructure directly?

```js
const { debounceRate, dedupKeyFn, onError, manualFetch = false } = initialOptions;
```

Reply:
Not with the hook as-is. The `fetchPageHandler` is wrapped in a `useCallback` that has `dedupKeyFn` in the dependency array. I initially did try to destructure directly but it led to a loop since the `dedupKeyFn` was a new instance each time it was called from the parent.

I figured instead of making calling components responsible for the memoization of the `dedupKeyFn` I would have the options be non-reactive and store them in local state (hence the _initial_ part of the parameter name). not as intuitive as I would like, but I thought it was better keeping the complexity inside and not forcing callers to `useCallback` themselves.

I can add a comment here explaining why.

Reply:
Yes, a comment sounds good.